incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -1026,49 +1106,95 @@ private void onKafkaData(
+                            final MqttWillSignalFW willSignal = sessionSignal.will();
+                            long deliverAt = willSignal.deliverAt();
+                            final String16FW willClientId = willSignal.clientId();
+
+                            if (deliverAt == MqttTime.UNKNOWN.value())
                             {
-                                deliverAt = supplyTime.getAsLong() + willSignal.delay();
+                                if (!instanceId.instanceId().equals(willSignal.instanceId()))
+                                {
+                                    deliverAt = supplyTime.getAsLong() + willSignal.delay();
+                                }
+                                else
+                                {
+                                    break reactToSignal;
+                                }
                             }
-                            else
+
+                            KafkaFetchWillStream willFetcher =
+                                new KafkaFetchWillStream(originId, routedId, this, sessionsTopic, willClientId,
+                                    willSignal.willId().asString(), willSignal.lifetimeId().asString(), deliverAt);
+                            willFetcher.doKafkaBegin(traceId, authorization, 0, willSignal.lifetimeId());
+                            willFetchers.put(new String16FW(willClientId.asString()), willFetcher);
+                            break;
+                        case MqttSessionSignalFW.KIND_EXPIRY:
+                            final MqttExpirySignalFW expirySignal = sessionSignal.expiry();
+                            long expireAt = expirySignal.expireAt();
+                            final String16FW expiryClientId = expirySignal.clientId();
+
+                            if (expireAt == MqttTime.UNKNOWN.value())
                             {
-                                break fetchWill;
+                                if (!instanceId.instanceId().equals(expirySignal.instanceId()))
+                                {
+                                    expireAt = supplyTime.getAsLong() + expirySignal.delay();
+                                }
+                                else
+                                {
+                                    break reactToSignal;
+                                }
                             }
-                        }
 
-                        KafkaFetchWillProxy willFetcher = new KafkaFetchWillProxy(originId, routedId, this, sessionsTopic,
-                            clientId, willSignal.willId().asString(), willSignal.lifetimeId().asString(), deliverAt);
-                        willFetcher.doKafkaBegin(traceId, authorization, 0, willSignal.lifetimeId());
-                        willFetchers.put(clientId, willFetcher);
+                            final int contextId = System.identityHashCode(expiryClientId.asString().intern());
```

Comment:
Let's avoid calling `String.intern()` on values specified externally, like clientIds.
Instead we can use a counter stored at the factory, or similar.